src/main/java/com/limechain/runtime/RuntimeBuilder.java

```diff
@@ -72,31 +160,48 @@ private byte[] zstDecompressIfNecessary(byte[] code) {
+        RuntimeVersion runtimeVersion = WasmSectionUtils.parseRuntimeVersionFromCustomSections(wasmBinary);
 
-        ScaleCodecReader reader = new ScaleCodecReader(data);
-        RuntimeVersionReader runtimeVersionReader = new RuntimeVersionReader();
-        return runtimeVersionReader.read(reader);
-    }
+        // If we couldn't get the data from the wasm custom sections,
+        // we must fall back to calling Core_version
+        if (runtimeVersion == null) {
+            log.log(Level.INFO, "Couldn't fetch runtime version from custom section, calling 'Core_version'.");
+            runtimeVersion = runtime.callCoreVersion();
+        }
 
-    static Imports getImports(Module module, Runtime runtime) {
-        ImportObject.MemoryImport memory = new ImportObject.MemoryImport("env", 22, false);
+        context.setRuntimeVersion(runtimeVersion);
+    }
 
-        ArrayList<ImportObject> objects = new ArrayList<>();
-        objects.addAll(StorageHostFunctions.getFunctions(runtime));
-        objects.addAll(ChildStorageHostFunctions.getFunctions(runtime));
-        objects.addAll(CryptoHostFunctions.getFunctions(runtime));
-        objects.addAll(HashingHostFunctions.getFunctions(runtime));
-        objects.addAll(OffchainHostFunctions.getFunctions(runtime));
-        objects.addAll(TrieHostFunctions.getFunctions(runtime));
-        objects.addAll(MiscellaneousHostFunctions.getFunctions(runtime));
-        objects.addAll(AllocatorHostFunctions.getFunctions(runtime));
-        objects.add(memory);
-
-        return Imports.from(objects, module);
+    /**
+     * A configuration holding all necessary dependencies for runtime calls.
+     *
+     * @param blockTrieAccessor used by storage related endpoints for accessing the trie storage for a block
+     * @param keyStore          used by crypto host functions to access secret keys
+     * @param offchainStorages  used by offchain host functions to access the different offchain storages
+     * @param isValidator       used by offchain host functions to determine
+     *                          whether the node is a validator (i.e. authoring node) or not
+     * @apiNote Explicitly passing `null`s for some of the dependencies is admissible in case you desire a partial
+     * runtime execution context (e.g. when you need a really lightweight runtime invocation).
+     * The implicit knowledge of whether a dependency will be used for the runtime invocations about to be performed,
+     * however, still remains a responsibility of the caller.
+     */
+    public record Config(
+        @Nullable
+        BlockTrieAccessor blockTrieAccessor,
+        @Nullable
+        KeyStore keyStore,
+        @Nullable
+        OffchainStorages offchainStorages,
+        boolean isValidator
+    ) {
+        public static final Config EMPTY = new Config(null, null, null, false);
     }
 
+    // TODO: Move this method somewhere else as it doesn't have to do with actually building a runtime instance
```

Comment:
This one's a big thorn for me. Has nothing to do with building a runtime instance; must be moved somewhere else. I've left this out for now as an out-of-scope change. 

That's essentially a static method. I would move it into `TrieDecoder` for example; or introduce a new utility class for that sole purpose; or move it closer to the user-site... not sure.

Reply:
After some looking into Murki's PR, `WarpSyncState` seems like the place where this method should live.

Reply:
Fixed in 00d07ab0ad7c371f92c6874200986cb4c61fbe42